Flink Delta Sink - refactor partition computer

pkubit-g opened the pull request:
- expose DeltaPartitionComputer instead of DeltaTablePartitionAssigner (bucket assginer)
- add RowDataPartitionComputer with utility class `DeltaPartitionComputer.forRowData`

scottsand-db commented:
javadoc for commit `42970c5970ed80c85ee8ea66bcb16035063a6e25`
[javaunidoc.zip](https://github.com/delta-io/connectors/files/8099310/javaunidoc.zip)

scottsand-db commented on flink-connector/README.md:
```diff
@@ -157,40 +157,26 @@ import java.util.LinkedHashMap;
+  public DataStream<RowData> createDeltaSink(DataStream<RowData> stream,
+                                             String deltaTablePath) {
+    List<String> partitionCols = Arrays.asList("surname");  
+    DeltaPartitionComputer<RowData> partitionComputer =
```
noob question, maybe I am missing something. remind me: why do we expose DeltaPartitionComputer?

Can't we have something like 
```java
BucketAssigner<RowData, String> bucketAssigner = myHelperFunction(ROW_TYPE, partitionCols);
myBuilder...withBucketAssigner(bucketAssigner)
```

Doesn't this remove the need for PartitionComputer?
```

pkubit-g commented on flink-connector/src/main/java/io/delta/flink/sink/DeltaPartitionComputer.java:
```diff
@@ -0,0 +1,139 @@
+    class DeltaRowDataPartitionComputer implements DeltaPartitionComputer<RowData> {
+
+        private final LinkedHashMap<String, String> staticPartitionSpec;
+        RowType rowType;
```
changed to private

pkubit-g commented on flink-connector/src/main/java/io/delta/flink/sink/DeltaPartitionComputer.java:
```diff
@@ -0,0 +1,139 @@
+        }
+
+        /**
+         * Creates instance of partition computer for {@link RowData}
```
Done & done & provided a code example with explanation

pkubit-g commented on flink-connector/src/main/java/io/delta/flink/sink/internal/DeltaTablePartitionAssigner.java:
```diff
@@ -40,10 +42,24 @@
+ *     // creates an instance of partition assigner that extracts partition values from
+ *     // {@link RowData} events
+ *     /////////////////////////////////////////////////////////////////////////////////
+ *     RowType rowType = ...;*
```
added by accident, corrected

pkubit-g commented on flink-connector/src/test/java/io/delta/flink/sink/internal/TestDeltaBucketAssignerInternal.java:
```diff
@@ -0,0 +1,323 @@
+/*
```
note: i haven't changed the whole file. Only renamed the main class and added two test cases for utility `forRowData` method

pkubit-g commented:
@scottsand-db thank you for the review.
I have corrected your remarks. Also, following your suggestions/question about the need for exposing the `DeltaPartitionComputer` i have:
- make the `DeltaPartitionComputer` internal
- renamed `DeltaTablePartitionAssigner` -> `DeltaBucketAssignerInternal` (and was already in the `internal` package)
- created `DeltaBucketAssigner` which extends `DeltaBucketAssignerInternal` and provides only two public utlity `forRowData` methods
- removed `DeltaSinkBuilder.withPartitionComputer` as now users will only need to call `DeltaBucketAssigner.forRowData(...)` to get the bucket assigner and then `DeltaSinkBuilder.withBucketAssigner` 

Following above changes i believe that some of your questions/suggestions become obsolete (those i haven't marked with thumb up emoi) so please do a double check on those changes and let me know if the current state lgty

pkubit-g commented:
javadoc for commit `265048da836139c6bb23e66dd8519d3263969875`
[javaunidoc.zip](https://github.com/delta-io/connectors/files/8128233/javaunidoc.zip)

scottsand-db commented:
javadoc for commit [a000985](https://github.com/delta-io/connectors/pull/281/commits/a0009857467b1f7bc4447bf890fb66bb06775ca5)
[javaunidoc.zip](https://github.com/delta-io/connectors/files/8155560/javaunidoc.zip)

tdas commented on examples/flink-example/src/main/java/io/delta/flink/example/sink/DeltaSinkPartitionedTableExample.java:
```diff
@@ -30,27 +34,15 @@ public static void main(String[] args) throws Exception {
-            partitionSpec.put("f3", Integer.toString(f3));
-            return partitionSpec;
-        }
+        return deltaSinkBuilder.build();
```
the code became so much simpler!!

tdas commented on flink-connector/src/main/java/io/delta/flink/sink/internal/DeltaBucketAssigner.java:
```diff
@@ -119,31 +106,6 @@ public String getBucketId(T element, Context context) {
-         */
-        LinkedHashMap<String, String> generatePartitionValues(
-            T element, Context context);
+        return "DeltaBucketAssignerInternal";
```
why Internal?

pkubit-g replied:
corrected

tdas commented on flink-connector/src/main/java/io/delta/flink/sink/DeltaSinkRowDataBuilder.java:
```diff
@@ -0,0 +1,167 @@
+ * configure additional behaviour (like merging of the schema or setting partition columns) and then
+ * build the sink.
+ */
+public class DeltaSinkRowDataBuilder {
```
I prefer the name to end with "DeltaSinkBuilder" because DeltaSink is  what you are building. So i recommend `RowDataDeltaSinkBuilder`

tdas commented on flink-connector/src/test/java/io/delta/flink/sink/internal/TestDeltaBucketAssignerInternal.java:
```diff
@@ -0,0 +1,262 @@
+import org.junit.Test;
+import static org.junit.Assert.assertEquals;
+
+public class TestDeltaBucketAssignerInternal {
```
its not internal any more .. isnt it?

pkubit-g commented on flink-connector/src/main/java/io/delta/flink/sink/internal/DeltaBucketAssigner.java:
```diff
@@ -60,48 +65,30 @@
  * @param <T> The type of input elements.
  */
-public class DeltaTablePartitionAssigner<T> implements BucketAssigner<T, String> {
+public class DeltaBucketAssigner<T> implements BucketAssigner<T, String> {
```
tbh i am not sure what are you referring to.
`DeltaBucketAssigner<T>` is a complete implementation that inherits from Flink's `BucketAssigner` which in order to work needs only defined mapping function between incoming event and its target partition path and this function is defined by providing implementation for `DeltaPartitionComputer<T>` interface that consists only from one function: `LinkedHashMap<String, String> generatePartitionValues`.

The other solution i see is to make `DeltaBucketAssigner` abstract and leave `LinkedHashMap<String, String> generatePartitionValues` unimplemented but to me the current composition is cleaner (with complete implementation for `DeltaBucketAssigner` and one functional interface).

scottsand-db replied:
TD's point is we don't need that functionality implementation defined in `DeltaPartitionComputer`.
This is just another class + another interface + another level of indirection.

Since both of these classes are internal, we can just move the `generatePartitionValues` method from the computer into the assigner, and then delete the computer.

Also then instantiate a concrete `DeltaRowDataBucketAssigner`.

Currently:
```java
// DeltaBucketAssigner.java
public String getBucketId(T element, Context context) {
        LinkedHashMap<String, String> partitionValues =
            this.partitionComputer.generatePartitionValues(element, context);
        return PartitionPathUtils.generatePartitionPath(partitionValues);
    }
    
// DeltaPartitionComputer.java
@override
public LinkedHashMap<String, String> generatePartitionValues(...) ....
```

TD's suggestion
```java
// DeltaBucketAssigner.java
public String getBucketId(T element, Context context) {
        LinkedHashMap<String, String> partitionValues =
            generatePartitionValues(element, context);
        return PartitionPathUtils.generatePartitionPath(partitionValues);
    }

public LinkedHashMap<String, String> generatePartitionValues(...)

// DeltaPartitionComputer.java has been DELETED
```

@tdas correct?

scottsand-db replied:
posting for visibility (as messaged on slack). do not block on this. this is all internal and we can refactor later.
